File: gamelogic/src/main/java/com/alta_v2/game/gamelogic/stage/StageManagerImpl.java
package com.alta_v2.game.gamelogic.stage;

import com.alta_v2.facade.coreApi.ScreenCoreApi;
import com.alta_v2.game.gamelogic.data.map.MapModel;
import com.alta_v2.game.gamelogic.data.npc.NpcModel;
import com.alta_v2.game.gamelogic.stage.event.ChangeMapStageEvent;
import com.alta_v2.game.gamelogic.stage.event.ChangeMenuStageEvent;
import com.alta_v2.game.gamelogic.stage.event.ChangeStageEvent;
import com.alta_v2.game.utils.ChangeScreenResult;
import com.alta_v2.mediator.serde.ActionListener;
import com.alta_v2.model.NpcDefinitionModel;
import com.google.inject.Inject;
import lombok.extern.log4j.Log4j2;

import java.util.List;
import java.util.stream.Collectors;

@Log4j2
public class StageManagerImpl implements StageManager {

    private final StageFactory stageFactory;
    private final ScreenCoreApi screenCoreApi;

    private Stage currentStage;

    @Inject
    public StageManagerImpl(StageFactory stageFactory, ScreenCoreApi screenCoreApi) {
        this.stageFactory = stageFactory;
        this.screenCoreApi = screenCoreApi;
        this.currentStage = this.createMenuStage();
    }

    /**
     * {@inheritDoc}
     */
    @Override
    public ActionListener getCurrentListener() {
        return this.currentStage;
    }

    private void onMapScreenChange(ChangeStageEvent data) {
        try {
            ChangeMapStageEvent event = ChangeStageEvent.resolve(data, ChangeMapStageEvent.class);
            currentStage.destroy();
            currentStage = createMenuStage();
            ChangeScreenResult loadResult = screenCoreApi.loadMenuScreen(event.getMapDefinition());
            if (loadResult != null) {
                loadResult.thenRun(currentStage::onStageLoaded);
            }
        } catch (NullPointerException | ClassCastException e) {
            log.error("Failed to change map screen", e);
        }
    }

    private void onMenuScreenChange(ChangeStageEvent data) {
        try {
            ChangeMenuStageEvent event = ChangeStageEvent.resolve(data, ChangeMenuStageEvent.class);
            currentStage.destroy();
            currentStage = createMapStage(
                    event.getMapDefinition().getDisplayName(), event.getMapDefinition().getNpcList()
            );
            ChangeScreenResult loadResult = screenCoreApi.loadTiledMapScreen(event.getMapDefinition());
            if (loadResult != null) {
                loadResult.thenRun(currentStage::onStageLoaded);
            }
        } catch (NullPointerException | ClassCastException e) {
            log.error("Failed to change menu screen", e);
        }
    }

    private void onMapChange(Void data) {
        log.info("Not implemented yet");
    }

    private Stage createMenuStage() {
        Stage stage = this.stageFactory.createMenuStage();
        stage.subscribeToChangeScreen(this::onMenuScreenChange);
        stage.subscribeToChangeMap(this::onMapChange);
        return stage;
    }

    private Stage createMapStage(String mapDisplayName, List<NpcDefinitionModel> definitionNpcList) {
        List<NpcModel> npcModels = definitionNpcList.stream()
                .map(npcDefinition ->
                        NpcModel.builder()
                                .id(npcDefinition.getId())
                                .localX(npcDefinition.getX())
                                .localY(npcDefinition.getY())
                                .repeatMovementInterval(npcDefinition.getRepeatMovementInterval())
                                .build()
                )
                .collect(Collectors.toList());

        MapModel mapModel = MapModel.builder().displayName(mapDisplayName).build();

        Stage stage = this.stageFactory.createMapStage(mapModel, npcModels);
        stage.subscribeToChangeScreen(this::onMapScreenChange);
        stage.subscribeToChangeMap(this::onMapChange);
        return stage;
    }
}

File: gamelogic/src/main/kotlin/com/alta_v2/game/gamelogic/stage/StageManagerImpl.kt
package com.alta_v2.game.gamelogic.stage

import com.alta_v2.exception.ChangeScreenException
import com.alta_v2.game.gamelogic.data.map.MapModel
import com.alta_v2.game.gamelogic.data.npc.toNpcModels
import com.alta_v2.game.gamelogic.domain.screen.ScreenProcessor
import com.alta_v2.game.gamelogic.stage.event.*
import com.alta_v2.game.utils.ChangeScreenResult
import com.alta_v2.mediator.serde.ActionListener
import com.alta_v2.model.NpcDefinitionModel
import com.google.inject.Inject
import mu.KotlinLogging

class StageManagerImpl @Inject constructor(private val stageFactory: StageFactory,
                                           private val screenProcessor: ScreenProcessor) : StageManager {

    private val log = KotlinLogging.logger {  }

    private var currentStage: Stage
    private var currentResult: ChangeScreenResult? = null

    init {
        currentStage = createMenuStage()
    }

    override fun getCurrentListener(): ActionListener = currentStage

    private fun onMapChange(data: Void) = log.info("Not implemented yet")

    private fun onMapScreenChange(data: ChangeStageEvent) {
        if (currentResult != null && !currentResult?.isDone!!) {
            log.warn("Failed to change map screen since another changing in progress")
            return
        }

        try {
            val event = data as ChangeMapStageEvent
            currentStage.destroy()
            currentStage = createMenuStage()

            currentResult = screenProcessor.showMenuScreen(event.mapDefinition)
            currentResult?.thenRun { currentStage.onStageLoaded() }
        } catch (e: ChangeScreenException) {
            log.error("Failed to change map screen", e)
        }
    }

    private fun onMenuScreenChange(data: ChangeStageEvent) {
        if (currentResult != null && !currentResult?.isDone!!) {
            log.warn("Failed to change map screen since another changing in progress")
            return
        }

        try {
            val event = data as ChangeMenuStageEvent
            currentStage.destroy()
            currentStage = createMapStage(
                    event.mapDefinition.displayName, event.mapDefinition.npcList
            )
            currentResult = screenProcessor.showMapScreen(event.mapDefinition)
            currentResult?.thenRun { currentStage.onStageLoaded() }
        } catch (e: ChangeScreenException) {
            log.error("Failed to change menu screen", e)
        }
    }

    private fun createMapStage(mapDisplayName: String, definitionNpcList: List<NpcDefinitionModel>): Stage {
        val stage: Stage = stageFactory.createMapStage(MapModel(mapDisplayName), definitionNpcList.toNpcModels())
        stage.subscribeToChangeScreen(createChangeScreenHandler(::onMapScreenChange))
        stage.subscribeToChangeMap(createChangeMapEventHandler(::onMapChange))
        return stage
    }

    private fun createMenuStage(): Stage {
        val stage: Stage = stageFactory.createMenuStage()
        stage.subscribeToChangeScreen(createChangeScreenHandler(::onMenuScreenChange))
        stage.subscribeToChangeMap(createChangeMapEventHandler(::onMapChange))
        return stage
    }

    private fun createChangeScreenHandler(handler: (data: ChangeStageEvent) -> Unit) =
            object : ChangeScreenHandler { override fun handle(data: ChangeStageEvent) = handler(data) }

    private fun createChangeMapEventHandler(handler: (data: Void) -> Unit) =
            object : ChangeMapEventHandler { override fun handle(data: Void) = handler(data) }
}